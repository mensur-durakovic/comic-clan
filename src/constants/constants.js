// grouping categories
export const YEAR = "Year";
export const WRITER = "Writer";
export const ARTIST = "Artist";
export const OWNER = "Owner";
export const RANDOM = "Random";
export const GROUPING_CATEGORIES = [YEAR, WRITER, ARTIST, OWNER, RANDOM];

// website texts
export const SEARCH_PLACEHOLDER = "Search by book name";
export const BACK_TEXT = "Back to collection";
export const OWNED_BY = "Owned By";
export const PUBLICATION = "Publication";
export const LOADING_TEXT = "Loading";

// api related stuff
export const COMMIC_CLAN_BASE_URL = "https://comicclan.vett.io/comics";

// text messages
export const NO_RESULTS_TEXT = "Ooops, looks like that comic doesn't exist! :(";
export const RANDOM_BOOKS_TITLE = "Other random books";

// error messages
export const ERROR_FETCH_BOOKS_TEXT = "Ooops, an error occurred!";
export const ERROR_SELECT_BOOK_TEXT = "Ooops, looks like that comic doesn't exist! Feel free to take a look at comics below :)";
export const ERROR_404_TEXT = "404 - page does not exist!";