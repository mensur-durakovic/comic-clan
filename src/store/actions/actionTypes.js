export const GET_COMMIC_BOOKS_FAILED = 'GET_COMMIC_BOOKS_FAILED';
export const SET_COMMIC_BOOKS = 'SET_COMMIC_BOOKS';
export const INIT_COMMIC_BOOKS = 'INIT_COMMIC_BOOKS';
export const SET_LOADING = 'SET_LOADING';
export const SELECT_BOOK = 'SELECT_BOOK';
export const SET_LOADING_BOOK = 'SET_LOADING_BOOK';

export const SEARCH = 'SEARCH';
export const SET_SEARCH_TERM = 'SET_SEARCH_TERM';
export const SET_SEARCH_RESULTS = 'SET_SEARCH_RESULTS';
export const CHANGE_CATEGORY = 'CHANGE_CATEGORY';
