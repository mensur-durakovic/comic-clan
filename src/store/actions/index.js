export {
    search,
    setSearchTerm,
    setSearchResults,
    initCommicBooks,
    setLoading,
    setLoadingBook,
    setCommicBooks,
    getCommicBooksFailed,
    changeCategory,
    selectBook,
} from './commicClan';
