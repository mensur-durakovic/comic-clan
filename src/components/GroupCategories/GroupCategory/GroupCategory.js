import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { useSelector, useDispatch } from "react-redux";
import * as actions from "../../../store/actions/index";

const GroupCategoryWrapper = styled.div`
  background-color: ${props => (props.isActive ? "#f15454" : "transparent")};
  box-shadow: ${props =>
    props.isActive ? "0px 2px 3px rgba(34, 34, 34, 0.6)" : "none"};
  color: ${props => (props.isActive ? "#FFFFFF" : "#777777")};
  border-radius: 20px;
  width: 75px;
  height: 35px;
  font-weight: bold;
  font-size: 16px;
  line-height: 35px;
  margin-right: 20px;
  margin-bottom: 20px;
  text-align: center;
  cursor: pointer;

`;

const GroupCategory = React.memo(props => {
  const activeCategory = useSelector(state => state.activeCategory);
  const dispatch = useDispatch();
  const { categoryName } = props;
  return (
    <GroupCategoryWrapper
      isActive={categoryName === activeCategory}
      onClick={() => dispatch(actions.changeCategory(categoryName))}
    >
      {categoryName}
    </GroupCategoryWrapper>
  );
});

GroupCategory.propTypes = {
  categoryName: PropTypes.string.isRequired
};

export default GroupCategory;
