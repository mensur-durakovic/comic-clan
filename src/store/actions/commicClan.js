import * as actionTypes from './actionTypes';

export const changeCategory = (newCategory) => {
    return {
        type: actionTypes.CHANGE_CATEGORY,
        newCategory: newCategory
    };
}

export const initCommicBooks = () => {
    return {
        type: actionTypes.INIT_COMMIC_BOOKS
    };
};

export const selectBook = (bookId) => {
    return {
        type: actionTypes.SELECT_BOOK,
        bookId: bookId,
    };
};

export const setLoading = () => {
    return {
        type: actionTypes.SET_LOADING
    };
};

export const setLoadingBook = () => {
    return {
        type: actionTypes.SET_LOADING_BOOK
    };
};
export const search = (searchTerm) => {
    return {
        type: actionTypes.SEARCH,
        searchTerm: searchTerm
    };
};

export const getCommicBooksFailed = () => {
    return {
        type: actionTypes.GET_COMMIC_BOOKS_FAILED
    };
};

export const setCommicBooks = ( commicBooks ) => {
    return {
        type: actionTypes.SET_COMMIC_BOOKS,
        commicBooks: commicBooks
    };
};

export const setSearchTerm = (searchTerm) => {
    return {
        type: actionTypes.SEARCH,
        searchTerm: searchTerm
    };
}

export const setSearchResults = (searchResults) => {
    return {
        type: actionTypes.SET_SEARCH_RESULTS,
        searchResults: searchResults
    };
}