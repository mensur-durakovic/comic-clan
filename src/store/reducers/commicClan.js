import * as actionTypes from "../actions/actionTypes";
import groupBy from "lodash/groupBy";
import {
  YEAR,
  WRITER,
  ARTIST,
  OWNER,
  RANDOM,
  GROUPING_CATEGORIES,
} from "../../constants/constants";

const initialState = {
  books: [],
  groupedBooks: {},
  searchTerm: "",
  activeCategory: YEAR,
  selectedBook: "",
  isLoading: false,
  isLoadingBook: false,
  errorFetchBooks: false,
  errorBookNotFound: false,
};

const groupByCategory = (books, category) => {
  switch (category) {
    case YEAR:
      return groupBy(books, (b) => b.year);
    case WRITER:
      return groupBy(books, (b) => b.writer);
    case ARTIST:
      return groupBy(books, (b) => b.artist);
    case OWNER:
      return groupBy(books, (b) => b.owner);
    case RANDOM:
      const randomNumber = Math.floor(
        Math.random() * GROUPING_CATEGORIES.length
      );
      return groupByCategory(books, GROUPING_CATEGORIES[randomNumber]);
    default:
      return groupBy(books, (b) => b.year);
  }
};

const setCommicBooks = (state, action) => {
  const groupedCommicBooks = groupByCategory(action.commicBooks, YEAR);
  return {
    ...state,
    books: action.commicBooks,
    groupedBooks: groupedCommicBooks,
    errorFetchBooks: false,
    errorBookNotFound: false,
    isLoading: false,
    isLoadingBook: false,
  };
};

const setSearchResults = (state, action) => {
  if (!action.searchResults.length) {
    return {
      ...state,
      books: [],
      groupedBooks: {},
      errorFetchBooks: false,
      errorBookNotFound: false,
      isLoading: false,
      isLoadingBook: false,
    };
  } else {
    const groupedCommicBooks = groupByCategory(
      action.searchResults,
      state.activeCategory
    );
    return {
      ...state,
      books: action.searchResults,
      groupedBooks: groupedCommicBooks,
      errorFetchBooks: false,
      errorBookNotFound: false,
      isLoading: false,
      isLoadingBook: false,
    };
  }
};

const getCommicBooksFailed = (state, action) => {
  return {
    ...state,
    errorFetchBooks: true,
    isLoading: false,
    isLoadingBook: false,
  };
};

const setLoading = (state, action) => {
  return { ...state, isLoading: true };
};

const setLoadingBook = (state, action) => {
  return { ...state, isLoadingBook: true };
};

const changeCategory = (state, action) => {
  const groupedCommicBooks = groupByCategory(state.books, action.newCategory);
  return {
    ...state,
    activeCategory: action.newCategory,
    groupedBooks: groupedCommicBooks,
  };
};

const setSearchTerm = (state, action) => {
  return { ...state, searchTerm: action.searchTerm };
};

const selectBook = (state, action) => {
  const book = state.books.find((b) => b.name.split(" #")[1] === action.bookId);

  return book
    ? {
        ...state,
        selectedBook: book,
        isLoadingBook: false,
        errorBookNotFound: false,
      }
    : {
        ...state,
        selectedBook: null,
        isLoadingBook: false,
        errorBookNotFound: true,
      };
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case actionTypes.SET_COMMIC_BOOKS:
      return setCommicBooks(state, action);
    case actionTypes.SELECT_BOOK:
      return selectBook(state, action);
    case actionTypes.SET_LOADING:
      return setLoading(state, action);
    case actionTypes.SET_LOADING_BOOK:
      return setLoadingBook(state, action);
    case actionTypes.GET_COMMIC_BOOKS_FAILED:
      return getCommicBooksFailed(state, action);
    case actionTypes.CHANGE_CATEGORY:
      return changeCategory(state, action);
    case actionTypes.SET_SEARCH_TERM:
      return setSearchTerm(state, action);
    case actionTypes.SET_SEARCH_RESULTS:
      return setSearchResults(state, action);
    default:
      return state;
  }
};

export default reducer;
