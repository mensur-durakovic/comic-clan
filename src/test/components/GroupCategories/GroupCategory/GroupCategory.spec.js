import React from "react";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import renderer from "react-test-renderer";
import {changeCategory} from '../../../../store/actions/commicClan';
import { ARTIST, YEAR } from "../../../../constants/constants";
import { commicBooks, groupedBooks } from "../../../testUtilities/constants";
import GroupCategory from "../../../../components/GroupCategories/GroupCategory/GroupCategory";

const mockStore = configureStore([]);

describe("GroupCategory component", () => {
  let store;
  let component;
  beforeEach(() => {
    store = mockStore({
      books: commicBooks,
      groupedBooks: groupedBooks,
      activeCategory: YEAR
    });
    component = renderer.create(
      <Provider store={store}>
        <GroupCategory categoryName={ARTIST} />
      </Provider>
    );
  });

  it("should render with given state from Redux store", () => {
    expect(component.toJSON()).toMatchSnapshot();
  });
  
/*   it('should dispatch an action on button click', () => {
    renderer.act(() => {
      component.onClick();
    });
    expect(store.dispatch).toHaveBeenCalledTimes(1);
    expect(store.dispatch).toHaveBeenCalledWith(
        changeCategory({ newCategory: ARTIST })
    );
  }); */
});
