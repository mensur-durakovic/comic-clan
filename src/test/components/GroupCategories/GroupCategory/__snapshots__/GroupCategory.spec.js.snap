// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GroupCategory component should render with given state from Redux store 1`] = `
<div
  className="sc-AxjAm hemMAk"
  onClick={[Function]}
>
  Artist
</div>
`;
